src/Serializer/AbstractItemNormalizer.php

```diff
@@ -488,4 +498,71 @@ protected function normalizeRelation(PropertyMetadata $propertyMetadata, $relate
 
         return $iri;
     }
+
+    /**
+     * Sets the object to populate if allowed to and not set already. If you pass identifiers, it will search for them,
+     * otherwise it falls back to the provided IdentifiersExtractor.
+     *
+     * @param $data
+     * @param string $class
+     * @param array  $context
+     * @param array  $identifiers
+     */
+    protected function setObjectToPopulate($data, $class, array &$context)
+    {
+        if (isset($context[self::OBJECT_TO_POPULATE]) || !\is_array($data)) {
+            return;
+        }
+
+        $updateAllowed = $context['api_allow_update'] ?? false;
+        $identifiers = $this->getIdentifiersForDenormalization($class);
+        $identifiersData = \array_intersect_key($data, array_flip($identifiers));
+
+        if (0 === \count($identifiersData)) {
+            return;
+        }
+
+        if (!$updateAllowed) {
+            throw new InvalidArgumentException('Update is not allowed for this operation.');
+        }
+
+        // TODO: use $this->iriConverter->getIriFromPlainIdentifier() once https://github.com/api-platform/core/pull/1837 is merged.
+        $iri = implode(';', $identifiersData);
+
+        try {
+            $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getItemFromIri($iri, $context + ['fetch_data' => true]);
+        } catch (InvalidArgumentException $e) {
+            // https://github.com/api-platform/core/issues/857
+            $identifiers = $this->identifiersExtractor->getIdentifiersFromResourceClass($class);
+            $identifiersData = \array_intersect_key($data, array_flip($identifiers));
+            if (0 === \count($identifiersData)) {
+                throw $e;
+            }
+
+            // TODO: use $this->iriConverter->getIriFromPlainIdentifier() once https://github.com/api-platform/core/pull/1837 is merged.
+            $context[self::OBJECT_TO_POPULATE] = $this->iriConverter->getItemFromIri(
+                sprintf(
+                    '%s/%s',
+                    $this->iriConverter->getIriFromResourceClass($context['resource_class']),
+                    implode(';', $identifiersData)
+                ),
+                $context + ['fetch_data' => true]
+            );
+        }
```

Comment:
Looks like we're using twice the same logic here, can't we avoid to repeat these?

Reply:
@soyuka if you have some ideas, please say. I can't see a very nice way atm.

Reply:
Can you shed some light on why this is needed? Looks like a fallback algorithm that should go into the iriconverter?

Reply:
I've added a comment a few lines up. it is a fix for https://github.com/api-platform/core/issues/857

Not sure I agree with the fix, but I have kept it as per the original as a behat feature tests it.